pkg/operator/status.go

```diff
@@ -44,6 +45,26 @@ func updateCondition(cr *imageregistryv1.Config, condtype string, condstate oper
 	}
 
 	cr.Status.Conditions = conditions
+	updateDegradedMetric(condtype, condstate)
+}
+
+// updateDegradedMetric updates prometheus metric that indicates if we are
+// operating in Degraded mode.
+func updateDegradedMetric(condtype string, condstate operatorapiv1.OperatorCondition) {
+	if condtype != operatorapiv1.OperatorStatusTypeDegraded {
+		return
+	}
+	if condstate.Status == operatorapiv1.ConditionFalse {
+		metrics.Degraded(false)
+		return
+	}
+	// StorageNotConfigured is not considered to be notified through
+	// Degraded metric, it has its own indication.
+	if condstate.Reason == "StorageNotConfigured" {
+		metrics.Degraded(false)
+		return
+	}
```

Comment:
This feels problematic, though I understand the rationale.

I think a better solution is to include a reason label for the degraded metric. We can then differentiate between `StorageNotConfigured` and other degraded reasons.

Reply:
The problem with the labels is that "Reasons" are mutually exclusive, in a given moment of time `T` only _one or none_ of the labels must be set to `true`. This premise tells that when the Operator is setting itself as degraded by reason `Y` it will have to clear reasons(labels) for `W` and `Z` as the operator could  previously be degraded by any of the latter. The same is valid when leaving the Degraded state, all Reasons should have their values cleared(set to `false`). That involves to know all reasons for Degraded and to clean all of others when setting one of them to `true`. Here `true` and `false` are represented by a `1` or a `0` on a Gauge metric.

Another possible solution considered was to use a Counter instead of a Gauge. That would make the alert dependent on the metric's `increase` instead of the metric's `current value`, on this case we could generate the alert using a query similar to:

```
increase(image_registry_operator_degraded{reason!="StorageNotConfigured"}[5m]) > 0
```

That would alert if there was an increase on any "Reason" different than `StorageNotConfigured` on the last 5 minutes. The problem with this approach is that it is tightly coupled with events coming in on the Operator, if there is no event there is no increase and eventually the alert could solve itself(even though still Degraded). This solution will also create a "constantly increasing" Degraded metric and that might seem counter-intuitive for someone looking at it: "Is it switching back and forth from and to Degraded at this ratio? What is going on?". Other than that such a metric AFAIK would not be feasible to properly measure the 48 hours of "StorageNotConfigured": the alert for this reason behaves differently from all the others as we should be able to create a promql that would alert only if the observed increase for Reason "StorageNotConfigured" is happening for 48 consecutive hours, if it stopped to increase on the last, lets say, 2 minutes the alert should be off. I don't know if such a query is possible.

To fill in the request on this comment I am going to implement the first alternative as for me it seems like the only real option here.

Reply:
After experimenting with different implementations I managed to find a way of making this work by using Counters instead of Gauges. With this approach we don't need the whole "clean up old Degraded reasons" magic and that, by itself, is a win in terms of complexity. PTAL.